### PR TITLE
Add which

### PR DESCRIPTION
**Pull request by mitchellwrosen**
#7

### CODE REVIEW

**Comment by Gabriel439** [comment]
Could you merge `master` again?  The build failure in CI is my fault and is an issue in `master` that I just fixed

File: src/Turtle/Prelude.hs [hunk 1]
```diff
@@ -1090,6 +1093,32 @@ time io = do
...
+          Just paths -> do
+            path <- select (Text.split (== ':') paths)
+            let path' = Filesystem.fromText path </> cmd
+            exists <- testfile path'
+            if exists
+                then do
+                    perms <- liftIO $
+                        getmod path'
+                          `catchIOError`
+                            \e ->
```
Comment (Gabriel439) [inline src/Turtle/Prelude.hs]: One minor suggestion to improve readability.  Perhaps factor out the handler as a named function, like this:

```haskell
...
then do
    let handler e =
            if isPermissionError e || isDoesNotExistError e
                then return Directory.emptyPermissions
                else throwIO e
    perms <- liftIO (getmod path' `catchIOError` handler)
    ...
```

The other reason that I like this trick is that you can add a type signature to `handler` to specify the type of the exception to catch (although it's not necessary in this example)

File: src/Turtle/Prelude.hs [hunk 2]
```diff
@@ -1090,6 +1093,32 @@ time io = do
 hostname :: MonadIO io => io Text
 hostname = liftIO (fmap Text.pack getHostName)
 
+-- | Show the full path of an executable file
+which :: MonadIO io => FilePath -> io (Maybe FilePath)
+which cmd = fold go Control.Foldl.head
```
Comment (Gabriel439) [inline src/Turtle/Prelude.hs]: I was wondering if it would be worth just returning `go` (i.e. change the result to `Shell FilePath`).  This would allow the user to retrieve all matching programs on the path.  For the common case they could wrap `which` in `fold which Control.Foldl.head`, but if they wanted even better error handling they could do something like:

```haskell
example = do
    paths <- fold go Control.Foldl.list
    case paths of
        []     -> ... -- Warn the user that no matches are found
        [path] -> ... -- Do something with the path
        _      -> ... -- Warn the user that there are multiple matches
```

I don't feel strongly about this, though

**Comment by Gabriel439** [comment]
Other than those comments this looks good to me

File: src/Turtle/Prelude.hs (hunk 1 above)
Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: That does look nicer, but it doesn't work when the action is a do block! How would you structure this?

```haskell
then do
    let handler e = ...
         action = do
             ....
             ....
    action `catcIOError` handler
```

(assuming that `let action = ...` isn't ideal)

File: src/Turtle/Prelude.hs (hunk 2 above)
Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: I don't feel strongly about it either, but the list behavior is closer to `which --all`, if that's worth anything.

File: src/Turtle/Prelude.hs [hunk 3]
```diff
@@ -1090,6 +1093,32 @@ time io = do
...
+  where
+    go :: Shell FilePath
+    go =
+      need "PATH" >>= \case
+          Nothing -> empty
+          Just paths -> do
+            path <- select (Text.split (== ':') paths)
+            let path' = Filesystem.fromText path </> cmd
+            exists <- testfile path'
+            if exists
```
Comment (int-index) [inline src/Turtle/Prelude.hs]: I believe this `if` can be replaced by `guard exists`.

File: src/Turtle/Prelude.hs [hunk 4]
```diff
@@ -1090,6 +1093,32 @@ time io = do
...
+                then do
+                    perms <- liftIO $
+                        getmod path'
+                          `catchIOError`
+                            \e ->
+                                if isPermissionError e || isDoesNotExistError e
+                                    then return Directory.emptyPermissions
+                                    else throwIO e
+                    if Directory.executable perms
```
Comment (int-index) [inline src/Turtle/Prelude.hs]: Likewise, this `if` can be replaced by `guard (Directory.executable perms)`.

File: src/Turtle/Prelude.hs (hunk 2 above)
Comment (int-index) [inline src/Turtle/Prelude.hs]: Sound like a good argument in favor of the list behavior.

File: src/Turtle/Prelude.hs (hunk 1 above)
Comment (Gabriel439) [inline src/Turtle/Prelude.hs]: If I'm reading the  code correctly, the `catchIOError` only scopes over the `getMod`, which is not in a `do` block.  However, the answer to your question is that I would write it exactly the way you did, (i.e. a separate `let` binding for `action`)

File: src/Turtle/Prelude.hs (hunk 2 above)
Comment (Gabriel439) [inline src/Turtle/Prelude.hs]: Oh, I didn't know about `which --all`?  Then how about adding the `Shell` version as `whichAll` and the `head` version as `which`?

File: src/Turtle/Prelude.hs (hunk 1 above)
Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: It is only scoping over the `getMod`, only because I couldn't make it look pretty the other way :)

No point in catching the error and returning empty permissions to inspect - could instead just catch the exception outside of the permission check and return empty there.

File: src/Turtle/Prelude.hs (hunk 2 above)
Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: @int-index did you mean what Gabriel said? Or did you mean that `which` should return a Shell?

Comment (int-index) [inline src/Turtle/Prelude.hs]: I meant `io [FilePath]` but I like @Gabriel439's proposal more.

**Comment by mitchellwrosen** [comment]
Done

File: src/Turtle/Prelude.hs [hunk 5]
```diff
@@ -1095,6 +1099,33 @@ time io = do
...
 hostname = liftIO (fmap Text.pack getHostName)
 
+-- | Show the full path of an executable file
+which :: MonadIO io => FilePath -> io (Maybe FilePath)
+which cmd = fold (whichAll cmd) Control.Foldl.head
+
+-- | Show all matching executables in PATH, not just the first
+whichAll :: FilePath -> Shell FilePath
+whichAll cmd =
+    need "PATH" >>= \case
```
Comment (int-index) [inline src/Turtle/Prelude.hs]: `Shell` handles pattern failures as `empty`, so this line can be written as `Just paths <- need "PATH"`, removing another indentation level.

Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: Sweet. Moment.

Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: Done

File: src/Turtle/Prelude.hs [hunk 6]
```diff
@@ -1095,6 +1098,31 @@ time io = do
...
+which cmd = fold (whichAll cmd) Control.Foldl.head
+
+-- | Show all matching executables in PATH, not just the first
+whichAll :: FilePath -> Shell FilePath
+whichAll cmd = do
+  Just paths <- need "PATH"
+  path <- select (Text.split (== ':') paths)
+  let path' = Filesystem.fromText path </> cmd
+
+  exists <- testfile path'
```
Comment (Gabriel439) [inline src/Turtle/Prelude.hs]: You can go a step further and apply the same pattern match failure trick here:

```haskell
True <- testfile path'
```

Then you don't need the `guard`

Comment (mitchellwrosen) [inline src/Turtle/Prelude.hs]: And the second `guard`?

```haskell
let handler :: IOError -> Permissions
     handler = ...

perms <- liftIO (getmod path' `catchIOError` handler)
guard (executable perms)
```
or
```haskell
let handler :: IOError -> Bool
     handler = ...

True <- liftIO (fmap executable (getmod path') `catchIOError` handler)
```

Comment (int-index) [inline src/Turtle/Prelude.hs]: The second `guard` is more readable as is, in my opinion.

**Comment by int-index** [comment]
The PR needs another rebase on `master` to fix the build failure on GHC 7.6.3 (which was due to `need` being unavailable).

**Comment by int-index** [comment]
Have you tested that the function works correctly in all cases? If yes, it is good to merge.

**Comment by mitchellwrosen** [comment]
To be honest I didn't do much testing, just poked at it with a stick in ghci. It's pretty simple, though: tokenize $PATH, see if there is an executable thing there. I'm sure it fails in weird corner cases, like if there's a directory in $PATH that is `-x` or something.

Is there any specific test you'd like to see?

**Comment by Gabriel439** [comment]
Note that this is basically the same behavior as the Unix `which` so it should be well-behaved.  See: https://opensource.apple.com/source/shell_cmds/shell_cmds-118/which/which.c

**Comment by Gabriel439** [comment]
Alright, this looks great to me!  I'll go ahead and squash and merge.  Thanks for the contribution! :)
